Adding a Helm chart to install in Kubernetes

handyman5 opened the pull request:
Hello,

I've created a Helm chart for installing Booklore in a Kubernetes cluster. I tried to make it robust and to use the same configuration as the docker compose setup, to make it easier to deal with.

If you are interested in this but would like it presented in a different form (such as a different directory or a separate repo), I'm happy to make whatever changes you like. I'll keep maintaining it on my end, since I intend to use the service myself.

Thanks for this, it seems like a really great tool!

handyman5 commented on example-chart/values.yaml:
```diff
@@ -0,0 +1,127 @@
+    port: http
+
+# This section is for setting up autoscaling more information can be found here: https://kubernetes.io/docs/concepts/workloads/autoscaling/
+autoscaling:
+  enabled: false
+  minReplicas: 1
+  maxReplicas: 100
+  targetCPUUtilizationPercentage: 80
+  # targetMemoryUtilizationPercentage: 80
```
Sure, this is all in the default config that `helm create` gives you. I can strip it out.

werwolfby replied:
Cool! I never new that there is `helm create` Nice! This explains all this similar charts, I thought that community find some common templates 😄

Anyway, I suggest to remove this one, to avoid any confusion.

handyman5 commented on example-chart/values.yaml:
```diff
@@ -0,0 +1,127 @@
+  auth:
+    database: booklore
+    username: booklore-user
+    existingSecret: booklore-mariadb-password
```
The intention is that, if you don't have an existing secret, it gets created for you and persisted across chart upgrades. That way you can safely reconfigure the chart without having to manually reapply the secret, and you don't have to create a secret first to use it (although it works fine if you do). Is there a different behavior you'd prefer?

handyman5 replied:
It's a little confusing and I'm happy to do this differently. The tricky bit is that you can't templatize the contents of `values.yaml`. I have a `Secret` template that creates essentially `{{ .Release.Name }}-mariadb-password`, and my assumption was that `booklore` was going to be the most likely release name when installing the chart. I get that it's unintuitive and I'm happy to change it, but I did some digging and I couldn't find another helm chart that handled this case in a way I liked better. If you're willing to have the installation instructions include "create a database secrete with `kubectl create secret <whatever>` then we can drop this and the secret template piece, and just have the pod rely on the secret we've asserted exists.

werwolfby replied:
I will probably could find some examples...

But essentially if you have this value specified, then no secrets should be created, and this secret should be used for `envFrom`, or just `env` with `secretRef` in the deployment.

Which I don't see here.

But part with creating the secret is completely correct. If this value is not specified, then secret should be created. The only problem that I see in current template is here: 

https://github.com/booklore-app/booklore/pull/1010/files#diff-56177bad038308ee7904c6feeb182cb46a3d9f564af8a65e913fc4a3639fdd8bR1-R2

```mustache
{{- $secretName := printf "%s-mariadb-password" .Release.Name }}
{{- $existing := lookup "v1" "Secret" .Release.Namespace $secretName }}
```

Which expects that secret already exists, and it doesn't use name (`existingSecret`) from values and uses `.Release.Name`.

And this part:

https://github.com/booklore-app/booklore/pull/1010/files#diff-53a1d4395431d6a12e2f5638cdeb4f0be6ea2708ffc417d6830fe21b45ae4ebcR74-R78

```yaml
            - name: DATABASE_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: {{ .Release.Name }}-mariadb-password
                  key: mariadb-password
```

Again doesn't use `existingSecret` values.

You pointed it correctly that it expects default name. My idea here if the value `existingSecret` specified, then it should be used and there is no need to generate template. If this value is not specified then generate a new secret and use it in `Deployment`.

handyman5 commented on example-chart/templates/pvc.yaml:
```diff
@@ -0,0 +1,29 @@
+    requests:
+      storage: {{ .Values.persistence.dataVolume.size }}
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  annotations:
+    helm.sh/resource-policy: keep
+  labels:
+    app.kubernetes.io/name: {{ include "booklore.fullname" . }}
+  name: {{ include "booklore.fullname" . }}-books
+spec:
+  accessModes:
+    - ReadWriteOnce
+  resources:
+    requests:
+      storage: {{ .Values.persistence.booksVolume.size }}
```
I'll fix this tomorrow.

handyman5 commented on example-chart/templates/deployment.yaml:
```diff
@@ -0,0 +1,120 @@
+              value: {{ .Values.mariadb.auth.username }}
+            - name: DATABASE_PASSWORD
+              valueFrom:
+                secretKeyRef:
+                  name: {{ .Release.Name }}-mariadb-password
+                  key: mariadb-password
```
This matches the name of the secret that the chart creates, which we expect to also match `values.yaml` but does not automatically do so. As I said above, I'm happy to change this.

handyman5 commented:
Ok, I think I addressed all of your comments. Thanks so much for the clear feedback and the good examples to work from!

* I removed the autoscaling and replicaCount bits
* It turns out that the mariadb helm chart made a secret anyway, so I removed mine. Now you'll either get that chart's secret (which the deployment will use) or the existingSecret you specify (which the deployment will also use)
* The persistence volumes can be enabled/disabled and support leveraging existing pvc claims where provided or creating new pvcs if not. There's no native NFS support, but you can mark both as `disabled` and then use `extraVolumes` and `extraVolumeMounts` to generate the appropriate config for your use case.

I _think_ this hits all the notes you gave. If there's anything I missed or anything else you want, do please let me know.

werwolfby commented on example-chart/templates/deployment.yaml:
```diff
@@ -0,0 +1,141 @@
+            {{ end }}
+        {{- end }}
+        {{- with .Values.extraVolumes }}
+        {{- toYaml . | nindent 2 }}
```
`nindent 2` - sounds like not enough, should be `8` I assume.

werwolfby reviewed (commented):
Thanks a lot for help and time for implement the chart.

It looks good to me, except this 2 questions for `nindent`, I'll give it a try soon.

handyman5 commented:
Good catches on the indents, I always screw up spacing in helm charts. I realized I had too many newlines in the `claimName`s too, so I fixed that as well. Thanks again for all the great feedback.

adityachandelgit commented:
@werwolfby I haven’t had a chance to fully review the PR. Have you had an opportunity to make a final decision? Are we good to merge it?

werwolfby commented:
@adityachandelgit I was trying to install chart myself, but cause it doesn't published and I'm want to have everything in gitops it didn't workout.

So I can't confirm that it works or not. But from code point of view eveything that I see sounds reasonable.

And my advice is to merge it, and let community try it. And fix it on the way if we will find any issues.
